src/Raven.Server/Documents/AI/AiIntegrationsController.cs

```diff
@@ -0,0 +1,166 @@
+
+#pragma warning disable SKEXP0001
+
+namespace Raven.Server.Documents.AI;
+
+public class AiIntegrationsController : IDisposable
+{
+    private readonly Dictionary<AiConnectionStringIdentifier, (AiConnectionString, ITextEmbeddingGenerationService)> _embeddingsGenerationServiceByConnectionStringIdentifier;
+
+    private Dictionary<EmbeddingsGenerationTaskIdentifier, AiConnectionStringIdentifier> _connectionStringsByTaskIdentifiers;
+    private Dictionary<EmbeddingsGenerationTaskIdentifier, EmbeddingsGenerationConfiguration> _embeddingsGenerationConfigurationByTaskIdentifiers;
+
+    private readonly RavenLogger _logger;
+    private readonly SemaphoreSlim _databaseRecordChangeLock = new(initialCount: 1, maxCount: 1);
+
+    public DocumentDatabase Database { get; }
+
+    public AiIntegrationsController(DocumentDatabase database)
+    {
+        Database = database;
+        _embeddingsGenerationServiceByConnectionStringIdentifier = new();
+        _embeddingsGenerationConfigurationByTaskIdentifiers = new();
+        _connectionStringsByTaskIdentifiers = new();
+        _logger = database.Loggers.GetLogger<AiIntegrationsController>();
+
+        var storage = new EmbeddingsStorage(database);
+        var cacher = new QueryEmbeddingsCacher(database, database.DatabaseShutdown);
+
+        Embeddings = new EmbeddingsController(this, storage, cacher);
+    }
+
+    public EmbeddingsController Embeddings { get; private set; }
+
+    public bool TryGetEmbeddingsGenerationConfiguration(EmbeddingsGenerationTaskIdentifier taskIdentifier, out EmbeddingsGenerationConfiguration configuration)
+    {
+        return _embeddingsGenerationConfigurationByTaskIdentifiers.TryGetValue(taskIdentifier, out configuration);
+    }
+
+    public bool TryGetConnectionStringIdByEmbeddingsGenerationTask(EmbeddingsGenerationTaskIdentifier taskIdentifier, out AiConnectionStringIdentifier connectionString)
+    {
+        return _connectionStringsByTaskIdentifiers.TryGetValue(taskIdentifier, out connectionString);
+    }
+
+    public async Task HandleDatabaseRecordChangeAsync(DatabaseRecord record)
+    {
+        if (record == null)
+            return;
+
+        await _databaseRecordChangeLock.WaitAsync();
+        try
+        {
+            var activeConnectionStringsIds = UpdateEmbeddingsGenerationConfigurationsAndConnectionStrings(record);
+            await UpdateEmbeddingGenerationServices(record, activeConnectionStringsIds);
+            UpdateQueryEmbeddingsCacherState(activeConnectionStringsIds);
+
+            if (_logger.IsInfoEnabled)
+                _logger.Info($"Updated Embeddings Generation configurations, cacher state, and services for {activeConnectionStringsIds.Count} active connection strings.");
+        }
+        finally
+        {
+            _databaseRecordChangeLock.Release();
+        }
+    }
+
+    private HashSet<AiConnectionStringIdentifier> UpdateEmbeddingsGenerationConfigurationsAndConnectionStrings(DatabaseRecord record)
```

Comment:
this method is incorrect, you cannot modify the Dictionary, even under lock, you might have reads during that time.
Either switch to Concurrent or create a new instance of the dictionary and substitute the field value